File: impl/python/agentnet_py/identity.py
from __future__ import annotations

from dataclasses import dataclass

from .cbor import CborMap, CborValue


class IdentityError(ValueError):
    pass


@dataclass
class IdentityRegisterPayload:
    agent_id: str
    pk_ed25519: bytes
    pk_x25519: bytes
    created: int


@dataclass
class IdentityRotatePayload:
    agent_id: str
    pk_ed25519: bytes
    pk_x25519: bytes
    ts: int


@dataclass
class CredentialRevokePayload:
    issuer: str
    credential_id_hash: bytes
    ts: int


def parse_identity_register_payload(value: CborValue) -> IdentityRegisterPayload:
    entries = _expect_map(value)
    pk_ed25519 = _expect_bytes(_get_required(entries, 1))
    pk_x25519 = _expect_bytes(_get_required(entries, 2))
    if len(pk_ed25519) != 32 or len(pk_x25519) != 32:
        raise IdentityError("invalid public key length")
    return IdentityRegisterPayload(
        agent_id=_expect_text(_get_required(entries, 0)),
        pk_ed25519=pk_ed25519,
        pk_x25519=pk_x25519,
        created=_expect_u64(_get_required(entries, 3)),
    )


def parse_identity_rotate_payload(value: CborValue) -> IdentityRotatePayload:
    entries = _expect_map(value)
    pk_ed25519 = _expect_bytes(_get_required(entries, 1))
    pk_x25519 = _expect_bytes(_get_required(entries, 2))
    if len(pk_ed25519) != 32 or len(pk_x25519) != 32:
        raise IdentityError("invalid public key length")
    return IdentityRotatePayload(
        agent_id=_expect_text(_get_required(entries, 0)),
        pk_ed25519=pk_ed25519,
        pk_x25519=pk_x25519,
        ts=_expect_u64(_get_required(entries, 3)),
    )


def parse_credential_revoke_payload(value: CborValue) -> CredentialRevokePayload:
    entries = _expect_map(value)
    cred_hash = _expect_bytes(_get_required(entries, 1))
    if len(cred_hash) != 32:
        raise IdentityError("credential_id_hash must be 32 bytes")
    return CredentialRevokePayload(
        issuer=_expect_text(_get_required(entries, 0)),
        credential_id_hash=cred_hash,
        ts=_expect_u64(_get_required(entries, 2)),
    )


def identity_register_payload_to_cbor(payload: IdentityRegisterPayload) -> CborValue:
    return CborMap([
        (0, payload.agent_id),
        (1, payload.pk_ed25519),
        (2, payload.pk_x25519),
        (3, payload.created),
    ])


def identity_rotate_payload_to_cbor(payload: IdentityRotatePayload) -> CborValue:
    return CborMap([
        (0, payload.agent_id),
        (1, payload.pk_ed25519),
        (2, payload.pk_x25519),
        (3, payload.ts),
    ])


def credential_revoke_payload_to_cbor(payload: CredentialRevokePayload) -> CborValue:
    return CborMap([
        (0, payload.issuer),
        (1, payload.credential_id_hash),
        (2, payload.ts),
    ])


def _expect_map(value: CborValue):
    if isinstance(value, CborMap):
        return value.entries
    raise IdentityError("expected map")


def _get_required(entries, key: int) -> CborValue:
    for k, v in entries:
        if isinstance(k, int) and k == key:
            return v
    raise IdentityError("missing required key")


def _expect_text(value: CborValue) -> str:
    if isinstance(value, str):
        return value
    raise IdentityError("expected text")


def _expect_bytes(value: CborValue) -> bytes:
    if isinstance(value, (bytes, bytearray)):
        return bytes(value)
    raise IdentityError("expected bytes")


def _expect_u64(value: CborValue) -> int:
    if isinstance(value, int) and value >= 0:
        return value
    raise IdentityError("expected unsigned")

File: impl/ts/src/identity.ts
import { CborMap, CborValue } from "./cbor.js";

export interface IdentityRegisterPayload {
  agentId: string;
  pkEd25519: Uint8Array;
  pkX25519: Uint8Array;
  created: number;
}

export interface IdentityRotatePayload {
  agentId: string;
  pkEd25519: Uint8Array;
  pkX25519: Uint8Array;
  ts: number;
}

export interface CredentialRevokePayload {
  issuer: string;
  credentialIdHash: Uint8Array;
  ts: number;
}

export function parseIdentityRegisterPayload(value: CborValue): IdentityRegisterPayload {
  const entries = expectMap(value);
  const pkEd = expectBytes(getRequired(entries, 1));
  const pkX = expectBytes(getRequired(entries, 2));
  if (pkEd.length !== 32 || pkX.length !== 32) {
    throw new Error("invalid public key length");
  }
  return {
    agentId: expectText(getRequired(entries, 0)),
    pkEd25519: pkEd,
    pkX25519: pkX,
    created: expectU64(getRequired(entries, 3)),
  };
}

export function parseIdentityRotatePayload(value: CborValue): IdentityRotatePayload {
  const entries = expectMap(value);
  const pkEd = expectBytes(getRequired(entries, 1));
  const pkX = expectBytes(getRequired(entries, 2));
  if (pkEd.length !== 32 || pkX.length !== 32) {
    throw new Error("invalid public key length");
  }
  return {
    agentId: expectText(getRequired(entries, 0)),
    pkEd25519: pkEd,
    pkX25519: pkX,
    ts: expectU64(getRequired(entries, 3)),
  };
}

export function parseCredentialRevokePayload(value: CborValue): CredentialRevokePayload {
  const entries = expectMap(value);
  const cred = expectBytes(getRequired(entries, 1));
  if (cred.length !== 32) throw new Error("credentialIdHash must be 32 bytes");
  return {
    issuer: expectText(getRequired(entries, 0)),
    credentialIdHash: cred,
    ts: expectU64(getRequired(entries, 2)),
  };
}

export function identityRegisterPayloadToCbor(payload: IdentityRegisterPayload): CborValue {
  return {
    entries: [
      [0, payload.agentId],
      [1, payload.pkEd25519],
      [2, payload.pkX25519],
      [3, payload.created],
    ],
  } satisfies CborMap;
}

export function identityRotatePayloadToCbor(payload: IdentityRotatePayload): CborValue {
  return {
    entries: [
      [0, payload.agentId],
      [1, payload.pkEd25519],
      [2, payload.pkX25519],
      [3, payload.ts],
    ],
  } satisfies CborMap;
}

export function credentialRevokePayloadToCbor(payload: CredentialRevokePayload): CborValue {
  return {
    entries: [
      [0, payload.issuer],
      [1, payload.credentialIdHash],
      [2, payload.ts],
    ],
  } satisfies CborMap;
}

function expectMap(value: CborValue): [CborValue, CborValue][] {
  if (typeof value === "object" && value !== null && "entries" in value) {
    return (value as CborMap).entries;
  }
  throw new Error("expected map");
}

function getRequired(entries: [CborValue, CborValue][], key: number): CborValue {
  for (const [k, v] of entries) {
    const idx = toNumber(k);
    if (typeof idx === "number" && idx === key) return v;
  }
  throw new Error("missing required key");
}

function expectText(value: CborValue): string {
  if (typeof value === "string") return value;
  throw new Error("expected text");
}

function expectBytes(value: CborValue): Uint8Array {
  if (value instanceof Uint8Array) return value;
  throw new Error("expected bytes");
}

function expectU64(value: CborValue): number {
  if (typeof value === "bigint") {
    const asNumber = Number(value);
    if (!Number.isSafeInteger(asNumber)) throw new Error("expected u64");
    return asNumber;
  }
  if (typeof value === "number" && Number.isSafeInteger(value) && value >= 0) return value;
  throw new Error("expected u64");
}

function toNumber(value: CborValue): number | null {
  if (typeof value === "number") return value;
  if (typeof value === "bigint") {
    const asNumber = Number(value);
    if (!Number.isSafeInteger(asNumber)) return null;
    return asNumber;
  }
  return null;
}
